### PR TITLE
improve AsyncSpan API: make #asyncFinish and #prepareForAsync can not invoke more than once

### PR DESCRIPTION
**Pull request by qxo**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.
improve AsyncSpan API: make #asyncFinish and #prepareForAsync can not invoke more than once
make the API more strong: make the plugin code bug fail-fast

### CODE REVIEW

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java [hunk 1]
```diff
@@ -322,18 +322,26 @@ public AbstractSpan start(long startTime) {
...
     @Override public AbstractSpan prepareForAsync() {
+        if (inAsyncMode) {
+            throw new RuntimeException("Span is already  in async mode, please do not  call '#prepareForAsync'  more than on the same span!");
+        }
         context = ContextManager.awaitFinishAsync(this);
         isInAsyncMode = true;
         return this;
     }
 
+    private  volatile boolean  asyncStop = false;
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/trace/AbstractTracingSpan.java]: Also, why the field is put here randomly?

**Comment by zhaoyuguang** [comment]
By learning your design, 
it seems that asyncStop and isInAsyncMode are very different things to do, 
isInAsyncMode is only viewed in the asyncFinish method, 
asyncStop has view and assignment in the asyncFinish method, and there is code logic in the middle, 
obviously It will still be executed multiple times.

**Comment by qxo** [comment]
sorry for my merge mistake

**Comment by wu-sheng** [comment]
Please make sure your submitted PR has been well discussed and have the clear goal. Open too many PR will make you too hard to follow.

**Pull request merged**